# Venice Astronomical Tide Data Cleaning
In this notebook, we'll be consolidate data on stronomical tides from different meteorological stations in Venice into a single tidy dataset. This notebook will explore data from the following sources:
- Data on Atronomical Tide Levels 2022 and 2023
- Data from the following stations in venice:
    - Stazione Burano
    - Stazione Diga Sud Lido
    - Stazione Misericordia
    - Stazione Palazzo Cavalli
    - Stazione Piattaforma PG
    - Stazione Piattaforma
    - Stazione Punta Salute (Canal Grande)
    - Stazione San Giorgio

Additional project descriptions and in-depth details on the data can be found on the `README` of this repository.


In [2]:
import warnings
import datetime
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
warnings.filterwarnings('once')
%matplotlib inline

## Astronomical Tides Data

### Astronomical Tides 2022

In [159]:
# Import astronomical tides data 2022
astro_tide_2022 = pd.read_json('astronomical_tide_2022.json')

astro_tide_2022.rename(columns={'data':'date','valore':'value'}, inplace=True) #Translate column names from Italian
astro_tide_2022['date'] = pd.to_datetime(astro_tide_2022.date) #Convert the date column into a datetime object

print(astro_tide_2022.head())
print(astro_tide_2022.tail())

                 date  value
0 2022-01-01 00:00:00     45
1 2022-01-01 00:05:00     44
2 2022-01-01 00:10:00     43
3 2022-01-01 00:15:00     42
4 2022-01-01 00:20:00     42
                      date  value
105115 2022-12-31 23:35:00     18
105116 2022-12-31 23:40:00     19
105117 2022-12-31 23:45:00     19
105118 2022-12-31 23:50:00     19
105119 2022-12-31 23:55:00     19


In [132]:
# Explore the astronomical tides data
astro_tide_2022.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105120 entries, 0 to 105119
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   date    105120 non-null  datetime64[ns]
 1   value   105120 non-null  int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 1.6 MB


In [133]:
# Convert the date column into a date object then explore the data summary
astro_tide_2022.describe()

value
count  105120.000000
mean       32.222108
std        24.179518
min       -34.000000
25%        15.000000
50%        34.000000
75%        50.000000
max        87.000000

The dataset contains **105,120 observations** across two columns, one with the **date** and the other with the **value** (astronomical tide value). No missing values in the dataset. The data is focused on 2022 with dates ranging from Jan 1st to Dec 31, 2022. The astro tides values across the year range from -34 to 87 with an avergae of 32.222.

### Astronomical Tides 2023

In [160]:
# Import astronomical tides data 2023
astro_tide_2023 = pd.read_json('astronomical_tide_2023.json')

astro_tide_2023.rename(columns={'data':'date','valore':'value'}, inplace=True) #Translate column names from Italian
astro_tide_2023['date'] = pd.to_datetime(astro_tide_2023.date) #Convert the date column into a datetime object

print(astro_tide_2023.head())
print(astro_tide_2023.tail())

                 date  value
0 2023-01-01 00:00:00     21
1 2023-01-01 00:05:00     22
2 2023-01-01 00:10:00     22
3 2023-01-01 00:15:00     22
4 2023-01-01 00:20:00     23
                      date  value
105115 2023-12-31 23:35:00     38
105116 2023-12-31 23:40:00     39
105117 2023-12-31 23:45:00     40
105118 2023-12-31 23:50:00     41
105119 2023-12-31 23:55:00     42


In [135]:
astro_tide_2023.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105120 entries, 0 to 105119
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   date    105120 non-null  datetime64[ns]
 1   value   105120 non-null  int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 1.6 MB


In [136]:
astro_tide_2023.describe()

value
count  105120.000000
mean       32.662072
std        24.391790
min       -33.000000
25%        15.000000
50%        34.000000
75%        51.000000
max        86.000000

The dataset is very similar in structure to `astro_tide_2022`. The dataset also contains **105,120 observations** across two columns, one with the **date** and the other with the **value** (astronomical tide value). No missing values in the dataset. The data is focused on 2023 with dates ranging from Jan 1st to Dec 31, 2023. The astro tides values across the year range from -33 to 86 with an avergae of 32.662.

### Data Consolidation
We'll now combine the two datasets on astronomical tides, `astro_tides_2022` and `astro_tides_2023` to create one consolidated dataset with dates ranging from January 1st 2022 to December 31st, 2023.

In [161]:
# Concatenate the data to remain with the two columns but consolidated rows
astro_tide = pd.concat([astro_tide_2022,astro_tide_2023], ignore_index=True)
print(astro_tide.head())
print(astro_tide.tail())

                 date  value
0 2022-01-01 00:00:00     45
1 2022-01-01 00:05:00     44
2 2022-01-01 00:10:00     43
3 2022-01-01 00:15:00     42
4 2022-01-01 00:20:00     42
                      date  value
210235 2023-12-31 23:35:00     38
210236 2023-12-31 23:40:00     39
210237 2023-12-31 23:45:00     40
210238 2023-12-31 23:50:00     41
210239 2023-12-31 23:55:00     42


In [138]:
# General Data Info on Conslidation
astro_tide.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 210240 entries, 0 to 210239
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   date    210240 non-null  datetime64[ns]
 1   value   210240 non-null  int64         
dtypes: datetime64[ns](1), int64(1)
memory usage: 3.2 MB


In [119]:
# Group data into hourly blocks with value maintained as the average of the hour
grouped = astro_tide.resample('H', on='date').value.mean()

df = pd.DataFrame()
df['date'] = np.array(grouped.index)
df['value'] = grouped.values

print(df.head())
print(df.tail())

                 date      value
0 2022-01-01 00:00:00  40.416667
1 2022-01-01 01:00:00  31.166667
2 2022-01-01 02:00:00  25.000000
3 2022-01-01 03:00:00  24.000000
4 2022-01-01 04:00:00  29.666667
                     date      value
17515 2023-12-31 19:00:00 -10.000000
17516 2023-12-31 20:00:00  -4.500000
17517 2023-12-31 21:00:00   6.666667
17518 2023-12-31 22:00:00  21.000000
17519 2023-12-31 23:00:00  35.916667


In [120]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17520 entries, 0 to 17519
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   date    17520 non-null  datetime64[ns]
 1   value   17520 non-null  float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 273.9 KB


We now have a consolidated dataset with data from Astronomical tides from 2022 to 2023. We have also reduced the time interval from 5 minute blocks to one hour block with value maintained as an average of the tide values across the hour.We'll use this as the anchor dataset where data from the different Venice stations will be consolidated into.

## Data from Stations

### Stazione Burano

In [162]:
burano_station = pd.read_csv('Stazione_Burano.csv', delimiter=';') #Import data from Burano Station
burano_station.head()


Data  Burano Liv.Idrom. medio 1m
0  2022-11-10 00:00:00                        0.66
1  2022-11-10 00:05:00                        0.67
2  2022-11-10 00:10:00                        0.67
3  2022-11-10 00:15:00                        0.67
4  2022-11-10 00:20:00                        0.67

In [163]:
burano_station.rename(columns={'Data':'date','Burano Liv.Idrom. medio 1m':'burano_tide_level'}, inplace=True) #Transalte column names to English
burano_station['date'] = pd.to_datetime(burano_station.date) #Convert the date column into a datetime object

burano_station.tail()

date  burano_tide_level
26491 2023-02-09 23:40:00               0.12
26492 2023-02-09 23:45:00               0.13
26493 2023-02-09 23:50:00               0.14
26494 2023-02-09 23:55:00               0.15
26495 2023-02-10 00:00:00               0.16

In [164]:
burano_station.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26496 entries, 0 to 26495
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   date               26496 non-null  datetime64[ns]
 1   burano_tide_level  26496 non-null  float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 414.1 KB


In [165]:
#Conslidate data to the main dataframe
astro_tide = astro_tide.merge(burano_station, how='inner', on='date')
astro_tide.head()

date  value  burano_tide_level
0 2022-11-10 00:00:00     54               0.66
1 2022-11-10 00:05:00     54               0.67
2 2022-11-10 00:10:00     53               0.67
3 2022-11-10 00:15:00     53               0.67
4 2022-11-10 00:20:00     52               0.67

In [166]:
astro_tide.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26496 entries, 0 to 26495
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   date               26496 non-null  datetime64[ns]
 1   value              26496 non-null  int64         
 2   burano_tide_level  26496 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 828.0 KB


#### Stazione Diga Sud Lido

In [167]:
diga_station = pd.read_csv('Stazione_DigaSudLido.csv', delimiter=';')
diga_station.head()

Data  D.S. Lido Liv.Idrom. medio 1m  \
0  2022-11-10 00:00:00                           0.65   
1  2022-11-10 00:05:00                           0.64   
2  2022-11-10 00:10:00                           0.64   
3  2022-11-10 00:15:00                           0.63   
4  2022-11-10 00:20:00                           0.62   

   Faro Diga Lido V.Vento med.10m  Faro Diga Lido V.Vento max  \
0                             3.8                         4.5   
1                             3.9                         4.5   
2                             3.6                         4.3   
3                             3.6                         4.5   
4                             3.7                         4.3   

   Faro Diga Lido D.Vento med. 10m  
0                               23  
1                               31  
2                               33  
3                               34  
4                               38

In [168]:
diga_station.rename(columns={'Data':'date','D.S. Lido Liv.Idrom. medio 1m':'diga_tide_level','Faro Diga Lido V.Vento med.10m':'diga_average_wind_speed','Faro Diga Lido V.Vento max':'diga_max_wind_speed','Faro Diga Lido D.Vento med. 10m':'diga_wind_direction'}, inplace=True) #Transalte column names to English
diga_station['date'] = pd.to_datetime(diga_station.date) #Convert the date column into a datetime object

diga_station.tail()

date  diga_tide_level  diga_average_wind_speed  \
26492 2023-02-09 23:40:00             0.32                      2.3   
26493 2023-02-09 23:45:00             0.33                      2.4   
26494 2023-02-09 23:50:00             0.33                      2.2   
26495 2023-02-09 23:55:00             0.33                      2.1   
26496 2023-02-10 00:00:00             0.34                      2.2   

       diga_max_wind_speed  diga_wind_direction  
26492                  2.9                  339  
26493                  2.8                  336  
26494                  2.4                  340  
26495                  2.8                  333  
26496                  2.7                  343

In [169]:
diga_station.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26497 entries, 0 to 26496
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   date                     26497 non-null  datetime64[ns]
 1   diga_tide_level          26497 non-null  float64       
 2   diga_average_wind_speed  26497 non-null  float64       
 3   diga_max_wind_speed      26497 non-null  float64       
 4   diga_wind_direction      26497 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(1)
memory usage: 1.0 MB


In [170]:
#Conslidate data to the main dataframe
astro_tide = astro_tide.merge(diga_station, how='inner', on='date')
astro_tide.head()

date  value  burano_tide_level  diga_tide_level  \
0 2022-11-10 00:00:00     54               0.66             0.65   
1 2022-11-10 00:05:00     54               0.67             0.64   
2 2022-11-10 00:10:00     53               0.67             0.64   
3 2022-11-10 00:15:00     53               0.67             0.63   
4 2022-11-10 00:20:00     52               0.67             0.62   

   diga_average_wind_speed  diga_max_wind_speed  diga_wind_direction  
0                      3.8                  4.5                   23  
1                      3.9                  4.5                   31  
2                      3.6                  4.3                   33  
3                      3.6                  4.5                   34  
4                      3.7                  4.3                   38

#### Stazione Misericordia

In [171]:
misericordia_station = pd.read_csv('Stazione_Misericordia.csv', delimiter=';')
misericordia_station.head()

Data  Misericordia Liv.Idrom. medio 1m  \
0  2022-11-10 00:00:00                              0.71   
1  2022-11-10 00:05:00                              0.71   
2  2022-11-10 00:10:00                              0.71   
3  2022-11-10 00:15:00                              0.71   
4  2022-11-10 00:20:00                              0.71   

   Misericordia D.Vento med. 10m  Misericordia V.Vento med.10m  \
0                             12                           3.0   
1                             10                           3.1   
2                             13                           3.2   
3                             11                           3.3   
4                             29                           3.3   

   Misericordia V.Vento max  Misericordia Onda: Alt. sign.  \
0                       3.8                           0.14   
1                       3.8                           0.12   
2                       3.9                           0.10   
3                       4.0                           0.10   
4                       3.8                           0.10   

   Misericordia Onda: Alt. max  
0                         0.20  
1                         0.22  
2                         0.18  
3                         0.16  
4                         0.18

#### Stazione Palazzo Cavalli

In [14]:
palazzo_station = pd.read_csv('Stazione_PalazzoCavalli.csv', delimiter=';')
palazzo_station.head()

Data  Palazzo Cavalli Pressione  \
0  2022-11-10 00:00:00                     1018.9   
1  2022-11-10 00:05:00                     1018.9   
2  2022-11-10 00:10:00                     1018.8   
3  2022-11-10 00:15:00                     1018.8   
4  2022-11-10 00:20:00                     1018.7   

   Palazzo Cavalli Temperatura  Palazzo Cavalli Umidita'  \
0                         14.3                        77   
1                         14.1                        78   
2                         14.2                        78   
3                         14.2                        78   
4                         14.2                        78   

   Palazzo Cavalli Radiazione solare  Palazzo Cavalli Pioggia 5m  
0                                0.0                         0.0  
1                                0.0                         0.0  
2                                0.0                         0.0  
3                                0.0                         0.0  
4                                0.0                         0.0

#### Stazione Piattaforma PG

In [16]:
pg_piattaforma_station = pd.read_csv('Stazione_Piattaforma_pg.csv', delimiter=';')
pg_piattaforma_station.head()

Data  Piattaforma CNR Pioggia 5m
0  2022-11-10 00:00:00                         0.0
1  2022-11-10 01:00:00                         0.0
2  2022-11-10 02:00:00                         0.0
3  2022-11-10 03:00:00                         0.0
4  2022-11-10 04:00:00                         0.0

#### Stazione Piattaforma

In [17]:
piattaforma_station = pd.read_csv('Stazione_Piattaforma.csv', delimiter=';')
piattaforma_station.head()

Data  Piattaforma CNR Liv.Idrom. medio 1m  \
0  2022-11-10 00:00:00                                 0.64   
1  2022-11-10 00:05:00                                 0.63   
2  2022-11-10 00:10:00                                 0.63   
3  2022-11-10 00:15:00                                 0.62   
4  2022-11-10 00:20:00                                 0.61   

   Piattaforma CNR D.Vento med. 10m  Piattaforma CNR V.Vento med.10m  \
0                              31.0                              6.0   
1                              34.0                              6.3   
2                              32.0                              6.6   
3                              36.0                              6.8   
4                              35.0                              6.8   

   Piattaforma CNR V.Vento max  Piattaforma CNR Onda: Alt. sign.  \
0                          7.2                              0.28   
1                          7.0                              0.28   
2                          7.6                              0.28   
3                          7.1                              0.26   
4                          7.8                              0.26   

   Piattaforma CNR Onda: Alt. max  Piattaforma CNR Pressione  \
0                            0.51                     1018.5   
1                            0.51                     1018.5   
2                            0.51                     1018.4   
3                            0.47                     1018.3   
4                            0.47                     1018.4   

   Piattaforma CNR Temperatura  Piattaforma CNR Temp. Acqua  \
0                         15.6                         18.5   
1                         15.5                         18.5   
2                         15.5                         18.7   
3                         15.4                         18.4   
4                         15.4                         18.6   

   Piattaforma CNR Umidita'  Piattaforma CNR Radiazione solare  \
0                      75.0                                NaN   
1                      75.0                                NaN   
2                      74.0                                NaN   
3                      74.0                                NaN   
4                      76.0                                NaN   

   Piattaforma CNR Pioggia 5m  
0                         0.0  
1                         0.0  
2                         0.0  
3                         0.0  
4                         0.0

#### Stazione Punta Salute (Canal Grande)

In [18]:
punta_station = pd.read_csv('Stazione_PuntaSalute_CanalGrande.csv', delimiter=';')
punta_station.head()

Data  P. Salute Liv.Idrom. medio 1m  P. Salute Temp. Acqua
0  2022-08-10 00:00:00                           0.32                   26.9
1  2022-08-10 00:05:00                           0.30                   27.8
2  2022-08-10 00:10:00                           0.27                   27.9
3  2022-08-10 00:15:00                           0.26                   26.9
4  2022-08-10 00:20:00                           0.24                   27.8

In [19]:
sangiorgio_station = pd.read_csv('Stazione_SanGiorgio.csv', delimiter=';')
sangiorgio_station.head()

Data  San Giorgio D.Vento med. 10m  \
0  2022-11-10 00:00:00                          18.0   
1  2022-11-10 00:05:00                          18.0   
2  2022-11-10 00:10:00                          22.0   
3  2022-11-10 00:15:00                          25.0   
4  2022-11-10 00:20:00                          23.0   

   San Giorgio V.Vento med.10m  San Giorgio V.Vento max  \
0                          4.0                      4.9   
1                          4.3                      4.9   
2                          4.5                      5.2   
3                          4.3                      5.4   
4                          4.6                      5.2   

   San Giorgio Temperatura  San Giorgio Umidita'  \
0                     14.0                  79.0   
1                     14.0                  79.0   
2                     14.0                  79.0   
3                     14.0                  79.0   
4                     13.9                  79.0   

   San Giorgio Radiazione solare  
0                            0.0  
1                            0.0  
2                            0.0  
3                            0.0  
4                            0.0

### Consolidated Meteorological Data
Below is the available consolidated data from meteoroligcal stations. We'll explore the datasets from the stations further as there is an overlap with this dataset to a degree.

In [94]:
meteo_data = pd.read_csv('Dati_Meteo.csv', delimiter=";")
print(meteo_data.columns)
meteo_data.head()

Index(['Data', 'Palazzo Cavalli Pressione', 'Palazzo Cavalli Temperatura',
       'Palazzo Cavalli Umidita'', 'San Giorgio Radiazione solare',
       'San Giorgio D.Vento med. 10m', 'San Giorgio V.Vento med.10m',
       'San Giorgio V.Vento max', 'P. Salute Temp. Acqua'],
      dtype='object')


Data  Palazzo Cavalli Pressione  \
0  2022-11-10 00:00:00                     1018.9   
1  2022-11-10 00:05:00                     1018.9   
2  2022-11-10 00:10:00                     1018.8   
3  2022-11-10 00:15:00                     1018.8   
4  2022-11-10 00:20:00                     1018.7   

   Palazzo Cavalli Temperatura  Palazzo Cavalli Umidita'  \
0                         14.3                      77.0   
1                         14.1                      78.0   
2                         14.2                      78.0   
3                         14.2                      78.0   
4                         14.2                      78.0   

   San Giorgio Radiazione solare  San Giorgio D.Vento med. 10m  \
0                            0.0                          18.0   
1                            0.0                          18.0   
2                            0.0                          22.0   
3                            0.0                          25.0   
4                            0.0                          23.0   

   San Giorgio V.Vento med.10m  San Giorgio V.Vento max  P. Salute Temp. Acqua  
0                          4.0                      4.9                   16.6  
1                          4.3                      4.9                   16.6  
2                          4.5                      5.2                   16.7  
3                          4.3                      5.4                   16.6  
4                          4.6                      5.2                   16.6